From 576a63f9c11eb0357e10fc851aab309bc6053e0e Mon Sep 17 00:00:00 2001
From: Gilles Dartiguelongue <eva@gentoo.org>
Date: Sun, 31 May 2009 12:47:09 +0200
Subject: [PATCH] Fix parallel make install setgid race condition.

Use install-data-exec-hook for data kind of files and install-exec-hook
for binary files.
---
 glines/Makefile.am     |   13 +++++++++----
 gnibbles/Makefile.am   |   10 +++++++---
 gnobots2/Makefile.am   |    8 ++++++--
 gnometris/Makefile.am  |    8 ++++++--
 gnomine/Makefile.am    |    8 ++++++--
 gnotravex/Makefile.am  |   12 ++++++++----
 gnotski/Makefile.am    |   12 ++++++++----
 gtali/Makefile.am      |   10 +++++++---
 mahjongg/Makefile.am   |   12 ++++++++----
 same-gnome/Makefile.am |   12 ++++++++----
 10 files changed, 73 insertions(+), 32 deletions(-)

diff --git a/glines/Makefile.am b/glines/Makefile.am
index 6fb9542..b75d20e 100644
--- a/glines/Makefile.am
+++ b/glines/Makefile.am
@@ -68,7 +68,7 @@ SCOREFILES = \
 	Large	\
 	$(NULL)
 
-install-schemas-local: $(schema_DATA)
+install-schemas-hook: $(schema_DATA)
 if GCONF_SCHEMAS_INSTALL
 	if test -z "$(DESTDIR)" ; then \
 		for p in $^ ; do \
@@ -77,7 +77,7 @@ if GCONF_SCHEMAS_INSTALL
 	fi
 endif
 
-install-scorefiles-local:
+install-scorefiles-hook:
 	-$(mkinstalldirs) $(DESTDIR)$(scoredir)
 	## Migrate to the new scoring system (with a link for dual-version
 	## installs).
@@ -90,6 +90,11 @@ install-scorefiles-local:
 	  chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/glines.$$i.scores; \
 	  chmod 664 $(DESTDIR)$(scoredir)/glines.$$i.scores; \
 	done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/glines && chmod 2555 $(DESTDIR)$(bindir)/glines ; fi
 
-install-data-local: install-schemas-local install-scorefiles-local
+install-data-hook: install-schemas-hook install-scorefiles-hook
+
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/glines && chmod 2555 $(DESTDIR)$(bindir)/glines ;\
+	fi
+
diff --git a/gnibbles/Makefile.am b/gnibbles/Makefile.am
index c75def3..71ddc0d 100644
--- a/gnibbles/Makefile.am
+++ b/gnibbles/Makefile.am
@@ -168,7 +168,7 @@ DISTCLEANFILES = $(desktop_DATA) $(schema_DATA)
 
 SCOREFILES = 4.0 3.0 2.0 1.0 4.1 3.1 2.1 1.1
 
-install-data-local: $(ggzclient_dsc_in_files:.dsc.in=.dsc)
+install-data-hook: $(ggzclient_dsc_in_files:.dsc.in=.dsc)
 if GCONF_SCHEMAS_INSTALL
 	GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) $(GCONFTOOL) --makefile-install-rule $(top_builddir)/gnibbles/$(schema_DATA)
 endif
@@ -181,9 +181,13 @@ endif
 	   chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/gnibbles.$$i.scores; \
 	   chmod 664 $(DESTDIR)$(scoredir)/gnibbles.$$i.scores; \
 	 done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gnibbles && chmod 2555 $(DESTDIR)$(bindir)/gnibbles ; fi
 
-uninstall-local:
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gnibbles && chmod 2555 $(DESTDIR)$(bindir)/gnibbles ;\
+	fi
+
+uninstall-hook:
 if GGZ_CLIENT
 	-$(GGZ_CONFIG) -D --remove --modfile=gnibbles-client.dsc
 endif
diff --git a/gnobots2/Makefile.am b/gnobots2/Makefile.am
index 377ad20..c60eb48 100644
--- a/gnobots2/Makefile.am
+++ b/gnobots2/Makefile.am
@@ -86,7 +86,7 @@ EXTRA_DIST = AUTHORS \
              $(config_DATA) \
              $(schema_in_files)
 
-install-data-local:
+install-data-hook:
 if GCONF_SCHEMAS_INSTALL
 	GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) $(GCONFTOOL) --makefile-install-rule $(top_builddir)/gnobots2/$(schema_DATA)
 endif
@@ -100,6 +100,10 @@ endif
 	    touch $$fname2; chmod 664 $$fname2; chown $(scores_user):$(scores_group) $$fname2; \
 	    touch $$fname3; chmod 664 $$fname3; chown $(scores_user):$(scores_group) $$fname3; \
 	    done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gnobots2 && chmod 2555 $(DESTDIR)$(bindir)/gnobots2 ; fi
+
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gnobots2 && chmod 2555 $(DESTDIR)$(bindir)/gnobots2 ;\
+	fi
 
 DISTCLEANFILES = $(Games_DATA) $(schema_DATA)
diff --git a/gnometris/Makefile.am b/gnometris/Makefile.am
index 14c72b0..e819e34 100644
--- a/gnometris/Makefile.am
+++ b/gnometris/Makefile.am
@@ -108,7 +108,7 @@ EXTRA_DIST = AUTHORS \
 	$(pixmap_DATA) \
 	$(schema_in_files)
 
-install-data-local:
+install-data-hook:
 if GCONF_SCHEMAS_INSTALL
 	GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) $(GCONFTOOL) --makefile-install-rule $(top_builddir)/gnometris/$(schema_DATA)
 endif
@@ -116,6 +116,10 @@ endif
 	-touch $(DESTDIR)$(scoredir)/gnometris.scores
 	-chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/gnometris.scores
 	-chmod 664 $(DESTDIR)$(scoredir)/gnometris.scores
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gnometris && chmod 2555 $(DESTDIR)$(bindir)/gnometris ; fi
+
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gnometris && chmod 2555 $(DESTDIR)$(bindir)/gnometris ;\
+	fi
 
 DISTCLEANFILES = $(Games_DATA) $(schema_DATA)
diff --git a/gnomine/Makefile.am b/gnomine/Makefile.am
index d11ce97..acb4077 100644
--- a/gnomine/Makefile.am
+++ b/gnomine/Makefile.am
@@ -54,7 +54,7 @@ Games_DATA = $(Games_in_files:.desktop.in.in=.desktop)
 
 SCOREFILES = Small Medium Large Custom
 
-install-data-local:
+install-data-hook:
 if GCONF_SCHEMAS_INSTALL
 	GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) $(GCONFTOOL) --makefile-install-rule $(top_builddir)/gnomine/$(schema_DATA)
 endif
@@ -68,6 +68,10 @@ endif
 	   chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/gnomine.$$i.scores; \
 	   chmod 664 $(DESTDIR)$(scoredir)/gnomine.$$i.scores; \
 	 done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gnomine && chmod 2555 $(DESTDIR)$(bindir)/gnomine ; fi
+
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gnomine && chmod 2555 $(DESTDIR)$(bindir)/gnomine ;\
+	fi
 
 DISTCLEANFILES = $(Games_DATA) $(schema_DATA)
diff --git a/gnotravex/Makefile.am b/gnotravex/Makefile.am
index 865f11c..f75a91e 100644
--- a/gnotravex/Makefile.am
+++ b/gnotravex/Makefile.am
@@ -55,7 +55,7 @@ EXTRA_DIST = \
 CLEANFILES = $(desktop_DATA) $(schema_DATA)
 DISTCLEANFILES = $(desktop_DATA) $(schema_DATA)
 
-install-schemas-local: $(schema_DATA)
+install-schemas-hook: $(schema_DATA)
 if GCONF_SCHEMAS_INSTALL
 	if test -z "$(DESTDIR)" ; then \
 		for p in $^ ; do \
@@ -66,13 +66,17 @@ endif
 
 SCOREFILES = 2 3 4 5 6
 
-install-scorefiles-local:
+install-scorefiles-hook:
 	-$(mkinstalldirs) $(DESTDIR)$(scoredir)
 	-for i in ${SCOREFILES} ; do \
 	   touch $(DESTDIR)$(scoredir)/gnotravex."$$i"x"$$i".scores; \
 	   chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/gnotravex."$$i"x"$$i".scores; \
 	   chmod 664 $(DESTDIR)$(scoredir)/gnotravex."$$i"x"$$i".scores; \
 	 done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gnotravex && chmod 2555 $(DESTDIR)$(bindir)/gnotravex ; fi
 
-install-data-local: install-schemas-local install-scorefiles-local
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gnotravex && chmod 2555 $(DESTDIR)$(bindir)/gnotravex ;\
+	fi
+
+install-data-hook: install-schemas-hook install-scorefiles-hook
diff --git a/gnotski/Makefile.am b/gnotski/Makefile.am
index 08efd17..3863937 100644
--- a/gnotski/Makefile.am
+++ b/gnotski/Makefile.am
@@ -62,7 +62,7 @@ EXTRA_DIST = \
 CLEANFILES = $(desktop_DATA) $(schema_DATA)
 DISTCLEANFILES = $(desktop_DATA) $(schema_DATA)
 
-install-schemas-local: $(schema_DATA)
+install-schemas-hook: $(schema_DATA)
 if GCONF_SCHEMAS_INSTALL
 	if test -z "$(DESTDIR)" ; then \
 		for p in $^ ; do \
@@ -71,16 +71,20 @@ if GCONF_SCHEMAS_INSTALL
 	fi
 endif
 
-install-scorefiles-local:
+install-scorefiles-hook:
 	-$(mkinstalldirs) $(DESTDIR)$(scoredir)
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gnotski && chmod 2555 $(DESTDIR)$(bindir)/gnotski ; fi
 	-for i in ${SCOREFILES} ; do \
 	  touch             $(DESTDIR)$(scoredir)/gnotski.$$i.scores; \
 	  chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/gnotski.$$i.scores; \
 	  chmod 664         $(DESTDIR)$(scoredir)/gnotski.$$i.scores; \
 	done
 
-install-data-local: install-schemas-local install-scorefiles-local
+install-data-hook: install-schemas-hook install-scorefiles-hook
+
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gnotski && chmod 2555 $(DESTDIR)$(bindir)/gnotski ;\
+	fi
 
 @INTLTOOL_SCHEMAS_RULE@
 @INTLTOOL_DESKTOP_RULE@
diff --git a/gtali/Makefile.am b/gtali/Makefile.am
index ca0ae91..ee6b0d5 100644
--- a/gtali/Makefile.am
+++ b/gtali/Makefile.am
@@ -53,7 +53,7 @@ Games_in_files = gtali.desktop.in.in
 Games_DATA = $(Games_in_files:.desktop.in.in=.desktop)
 @INTLTOOL_DESKTOP_RULE@
 
-install-data-local:
+install-data-hook:
 if GCONF_SCHEMAS_INSTALL
 	-GCONF_CONFIG_SOURCE=$(GCONF_SCHEMA_CONFIG_SOURCE) $(GCONFTOOL) --makefile-install-rule $(top_builddir)/gtali/$(schema_DATA)
 endif
@@ -64,7 +64,11 @@ endif
 	-chmod 664 $(DESTDIR)$(scoredir)/gtali.Regular.scores	
 	-touch $(DESTDIR)$(scoredir)/gtali.Colors.scores
 	-chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/gtali.Colors.scores
-	-chmod 664 $(DESTDIR)$(scoredir)/gtali.Colors.scores	
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/gtali && chmod 2555 $(DESTDIR)$(bindir)/gtali ; fi
+	-chmod 664 $(DESTDIR)$(scoredir)/gtali.Colors.scores
+
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/gtali && chmod 2555 $(DESTDIR)$(bindir)/gtali ;\
+	fi
 
 DISTCLEANFILES = $(Games_DATA) $(schema_DATA)
diff --git a/mahjongg/Makefile.am b/mahjongg/Makefile.am
index 8da5b46..d378de7 100644
--- a/mahjongg/Makefile.am
+++ b/mahjongg/Makefile.am
@@ -73,7 +73,7 @@ EXTRA_DIST = \
 CLEANFILES = $(desktop_DATA) $(schema_DATA)
 DISTCLEANFILES = $(desktop_DATA) $(schema_DATA)
 
-install-schemas-local: $(schema_DATA)
+install-schemas-hook: $(schema_DATA)
 if GCONF_SCHEMAS_INSTALL
 	if test -z "$(DESTDIR)" ; then \
 		for p in $^ ; do \
@@ -82,16 +82,20 @@ if GCONF_SCHEMAS_INSTALL
 	fi
 endif
 
-install-scorefiles-local:
+install-scorefiles-hook:
 	-$(mkinstalldirs) $(DESTDIR)$(scoredir)
 	-for i in easy difficult confounding pyramid tictactoe cloud dragon bridges ziggurat; do \
 		touch $(DESTDIR)$(scoredir)/mahjongg.$$i.scores; \
 		chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/mahjongg.$$i.scores; \
 		chmod 664 $(DESTDIR)$(scoredir)/mahjongg.$$i.scores; \
 	done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/mahjongg && chmod 2555 $(DESTDIR)$(bindir)/mahjongg ; fi
 
-install-data-local: install-schemas-local install-scorefiles-local
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/mahjongg && chmod 2555 $(DESTDIR)$(bindir)/mahjongg ;\
+	fi
+
+install-data-hook: install-schemas-hook install-scorefiles-hook
 
 @INTLTOOL_DESKTOP_RULE@
 @INTLTOOL_SCHEMAS_RULE@
diff --git a/same-gnome/Makefile.am b/same-gnome/Makefile.am
index e147957..406a3de 100644
--- a/same-gnome/Makefile.am
+++ b/same-gnome/Makefile.am
@@ -69,7 +69,7 @@ DISTCLEANFILES = $(desktop_DATA) $(schema_DATA)
 
 SCOREFILES = Small Medium Large
 
-install-schemas-local: $(schema_DATA)
+install-schemas-hook: $(schema_DATA)
 if GCONF_SCHEMAS_INSTALL
 	if test -z "$(DESTDIR)" ; then \
 		for p in $^ ; do \
@@ -78,7 +78,7 @@ if GCONF_SCHEMAS_INSTALL
 	fi
 endif
 
-install-scorefiles-local:
+install-scorefiles-hook:
 	-$(mkinstalldirs) $(DESTDIR)$(scoredir)
 	## Migrate to the new scoring system (with a link for dual-version
 	## installs).
@@ -91,9 +91,13 @@ install-scorefiles-local:
 	  chown $(scores_user):$(scores_group) $(DESTDIR)$(scoredir)/same-gnome.$$i.scores; \
 	  chmod 664 $(DESTDIR)$(scoredir)/same-gnome.$$i.scores; \
 	done
-	-if test "x$(setgid)" = "xtrue"; then chgrp $(scores_group) $(DESTDIR)$(bindir)/same-gnome && chmod 2555 $(DESTDIR)$(bindir)/same-gnome ; fi
 
-install-data-local: install-schemas-local install-scorefiles-local
+install-exec-hook:
+	-if test "x$(setgid)" = "xtrue"; then \
+	  chgrp $(scores_group) $(DESTDIR)$(bindir)/same-gnome && chmod 2555 $(DESTDIR)$(bindir)/same-gnome ;\
+	fi
+
+install-data-hook: install-schemas-hook install-scorefiles-hook
 
 @INTLTOOL_SCHEMAS_RULE@
 @INTLTOOL_DESKTOP_RULE@
-- 
1.6.3.1

